### PR TITLE
[Trainer] Add LoraLLMTrainer and Default Training Structure

### LINKED ISSUE #617: Implement Lora LLM Trainer Class
### Is your feature request related to a problem? Please describe

This issue is related to adding the first specific trainer class.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mmilosevicTT**
### Issue
Closes https://github.com/tenstorrent/tt-blacksmith/issues/617

### Problem description
This PR adds training through trainer class for lora llm models.

### What's Changed
- Added a generic `Trainer` base class (`blacksmith/tools/trainer/trainer.py`) that orchestrates the training loop: epochs, gradient accumulation, inline validation, optimizer setup, callback hooks, and device setup via `DeviceManager`.
- Integrated `ReproducibilityManager` into `Trainer.setup()` so seeds are set automatically from config (no manual seeding in scripts).
- Added `LoraLLMTrainer` strategy for parameter-efficient fine-tuning of causal LLMs (model/dataloader loading + forward/loss).
- Introduced a trainer config hierarchy: `TrainerConfig` (base) and `LoraLLMConfig`.
- Added an example runner (`examples/lora_llm/train.py`) plus SST-2 + Llama-3.2-1B configs for single-chip, tensor-parallel, and data-parallel setups.
- Updated the callback system to support the trainer lifecycle hooks.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: blacksmith/tools/trainer/configs/base.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    # Dataset settings
+    dataset_id: str = Field(default="dataset_id")
+
+    # Model settings
+    model_name: str = Field(default="model_name")
+    dtype: str = Field(default="dtype")
+
+    # Training hyperparameters
+    learning_rate: float = Field(default=2e-5, gt=0)
```
Comment (umalesTT) [inline blacksmith/tools/trainer/configs/base.py:24]: I'd add a scheduler field here

Comment (umalesTT) [reply blacksmith/tools/trainer/configs/base.py:24]: synced offline, it can be postponed

File: blacksmith/tools/trainer/configs/lora_llm.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+
+    # Model settings
+    max_length: int = Field(default=128, gt=0)
+
+    # Training hyperparameters
+    ignored_index: int = Field(default=-100)  # Label id ignored by the cross-entropy loss.
+
+    # LoRA setup
+    lora_r: int = Field(default=8, gt=0)
```
Comment (umalesTT) [inline blacksmith/tools/trainer/configs/lora_llm.py:21]: We should think of adding lora dropout

File: blacksmith/tools/trainer/examples/lora_llm/train.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from blacksmith.tools.trainer.configs import LoraLLMConfig
+from blacksmith.tools.trainer.strategies import LoraLLMTrainer
+
+if __name__ == "__main__":
+    default_config = Path(__file__).parent / "single_chip" / "llama_3_2_1b_sst2.yaml"
+    args = parse_cli_options(default_config=default_config)
+    config: LoraLLMConfig = generate_config(LoraLLMConfig, args.config)
+
+    trainer = LoraLLMTrainer()
+    trainer.setup(config)
```
Comment (pglusacTT) [inline blacksmith/tools/trainer/examples/lora_llm/train.py:16]: Why not as a part of the constructor?

File: blacksmith/tools/trainer/strategies/lora_llm_trainer.py [hunk 4]
```diff
@@ -1,10 +1,46 @@
...
+    Provides the strategy-specific pieces (model, dataloaders and the
+    forward/loss computation) on top of the generic training loop, optimizer,
+    gradient accumulation, validation and callback orchestration implemented in
+    :class:`~blacksmith.tools.trainer.trainer.Trainer`.
+    """
+
+    def _load_model(self) -> torch.nn.Module:
+        return get_model(self.config, self.device_manager.device, compile_model=True)
+
+    def _load_dataloaders(self) -> tuple:
```
Comment (pglusacTT) [inline blacksmith/tools/trainer/strategies/lora_llm_trainer.py]: Tuple of what?

Comment (mmilosevicTT) [reply blacksmith/tools/trainer/strategies/lora_llm_trainer.py]: Good catch. Added DataLoader type to it.

File: blacksmith/tools/trainer/examples/lora_llm/train.py (hunk 3 above)
Comment (mmilosevicTT) [reply blacksmith/tools/trainer/examples/lora_llm/train.py:16]: We discussed this offline. Since this trainer is meant to be general for different training methods, I would keep this as is for now and then we can improve on it when the need comes. I just do not want to overfit to one idea and then have a need to rework from beginning.

File: blacksmith/tools/trainer/trainer.py [hunk 5]
```diff
@@ -16,37 +23,231 @@ def __init__(
...
+        except Exception as exception:
+            self.callback_handler("on_error", exception)
+        finally:
+            self.callback_handler("on_train_end")
+
+    def validate(self) -> None:
+        """
+        Run the validation loop over ``self.val_dataloader``.
+
+        Reuses ``_forward`` under ``torch.no_grad`` (the forward pass is
```
Comment (umalesTT) [inline blacksmith/tools/trainer/trainer.py]: What do you mean by forward pass is identical to training?

**Review by umalesTT — APPROVED** [review]

File: blacksmith/tools/trainer/configs/base.py (hunk 1 above)
Comment (mmilosevicTT) [reply blacksmith/tools/trainer/configs/base.py:24]: https://github.com/tenstorrent/tt-blacksmith/issues/643

Opened an issue for that 😄

File: blacksmith/tools/trainer/configs/lora_llm.py (hunk 2 above)
Comment (mmilosevicTT) [reply blacksmith/tools/trainer/configs/lora_llm.py:21]: Yeah, can add it for this PR just leave it at no dropout.

Comment (mmilosevicTT) [reply blacksmith/tools/trainer/configs/lora_llm.py:21]: Actually this will make the PR explode, will add it as part of the refactor PR for configs.

File: blacksmith/tools/trainer/trainer.py (hunk 5 above)
Comment (mmilosevicTT) [reply blacksmith/tools/trainer/trainer.py]: Fixed

File: blacksmith/tools/trainer/configs/base.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    Holds fields that are common to all trainer configurations.
+    """
+
+    # Dataset settings
+    dataset_id: str = Field(default="dataset_id")
+
+    # Model settings
+    model_name: str = Field(default="model_name")
+    dtype: str = Field(default="dtype")
```
Comment (ndrakulicTT) [inline blacksmith/tools/trainer/configs/base.py:21]: Can we maybe change this to some enum or something similar to move away from using python `eval` ?

File: blacksmith/tools/trainer/configs/base.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    # Training hyperparameters
+    learning_rate: float = Field(default=2e-5, gt=0)
+    batch_size: int = Field(default=8, gt=0)
+    num_epochs: int = Field(default=1, gt=0)
+    optim: str = Field(default="adamw_torch")
+    weight_decay: float = Field(default=0.0, ge=0)
+    gradient_accumulation_steps: int = Field(default=1, ge=1)
+    gradient_checkpointing: bool = Field(default=False)
+    training_type: str = Field(default="lora")
```
Comment (ndrakulicTT) [inline blacksmith/tools/trainer/configs/base.py:31]: Do we need default here, it would maybe make more sense if we dont have default here but have it in the `LoRAConfig` overided with the default argument, that way child instances need to declare training_type, or it will need to be defined in `.yaml`. 
This also applies to most of other fields.

File: blacksmith/tools/trainer/callback.py [hunk 8]
```diff
@@ -31,10 +31,10 @@ def on_validation_start(self, trainer):
     def on_validation_batch_start(self, trainer, batch):
         pass
 
-    def on_validation_batch_end(self, trainer, batch):
+    def on_validation_batch_end(self, trainer, batch, loss):
         pass
 
-    def on_validation_end(self, trainer):
+    def on_validation_end(self, trainer, val_loss):
```
Comment (ndrakulicTT) [inline blacksmith/tools/trainer/callback.py:37]: Do we here, maybe want something like validation output or *args, **kwargs as we maybe want to have: examples, accuracy, other metrics?

File: blacksmith/tools/trainer/trainer.py [hunk 9]
```diff
@@ -16,37 +23,207 @@ def __init__(
...
+        # `on_train_start` callbacks need to guard against having a config.
+        if self.config is None:
+            return
+
+        grad_accumulation_steps = self.config.gradient_accumulation_steps
+        # TODO(mmilosevicTT): Temporary running loss for debug prints; remove once
+        # default logging callback lands. See https://github.com/tenstorrent/tt-blacksmith/issues/621
+        running_loss = 0.0
+        running_count = 0
+        try:
```
Comment (ndrakulicTT) [inline blacksmith/tools/trainer/trainer.py:94]: Would it be worth it to have some form of helper context menager, handle_exception. That would do
```python
@contextmanger
def handle_exception(f):
    try:
        yield f()
    except Exception as exception:
        self.callback_handler("on_error", trainer=self, exception=exception)
    finally:
        self.callback_handler("on_train_end")
```
That way we dont need whole training loop inside try catch
